test_conformance/commonfns/test_base.h

```diff
@@ -0,0 +1,282 @@
+    {}
+
+    cl_int Run() override;
+};
+
+//--------------------------------------------------------------------------
+
+struct DegreesTest : BaseFunctionTest
+{
+    DegreesTest(cl_device_id device, cl_context context, cl_command_queue queue,
+                int num_elems, const char *fn, bool vsp)
+        : BaseFunctionTest(device, context, queue, num_elems, fn, vsp)
+    {}
+
+    cl_int Run() override;
+};
+
+//--------------------------------------------------------------------------
+
+struct RadiansTest : BaseFunctionTest
+{
+    RadiansTest(cl_device_id device, cl_context context, cl_command_queue queue,
+                int num_elems, const char *fn, bool vsp)
+        : BaseFunctionTest(device, context, queue, num_elems, fn, vsp)
+    {}
+
+    cl_int Run() override;
+};
+
+//--------------------------------------------------------------------------
+
+struct SignTest : BaseFunctionTest
+{
+    SignTest(cl_device_id device, cl_context context, cl_command_queue queue,
+             int num_elems, const char *fn, bool vsp)
+        : BaseFunctionTest(device, context, queue, num_elems, fn, vsp)
+    {}
+
+    cl_int Run() override;
+};
+
+//--------------------------------------------------------------------------
+
+struct SmoothstepTest : BaseFunctionTest
+{
+    SmoothstepTest(cl_device_id device, cl_context context,
+                   cl_command_queue queue, int num_elems, const char *fn,
+                   bool vsp)
+        : BaseFunctionTest(device, context, queue, num_elems, fn, vsp)
+    {}
+
+    cl_int Run() override;
+};
+
+//--------------------------------------------------------------------------
+
+struct StepTest : BaseFunctionTest
+{
+    StepTest(cl_device_id device, cl_context context, cl_command_queue queue,
+             int num_elems, const char *fn, bool vsp)
+        : BaseFunctionTest(device, context, queue, num_elems, fn, vsp)
+    {}
+
+    cl_int Run() override;
+};
+
+//--------------------------------------------------------------------------
+
+struct MixTest : BaseFunctionTest
+{
+    MixTest(cl_device_id device, cl_context context, cl_command_queue queue,
+            int num_elems, const char *fn, bool vsp)
+        : BaseFunctionTest(device, context, queue, num_elems, fn, vsp)
+    {}
+
+    cl_int Run() override;
+};
+
+//--------------------------------------------------------------------------/
+
+template <typename... Args>
+std::string string_format(const std::string &format, Args... args)
+{
+    int sformat = std::snprintf(nullptr, 0, format.c_str(), args...) + 1;
+    if (sformat <= 0)
+        throw std::runtime_error("string_format: string processing error.");
+    auto format_size = static_cast<size_t>(sformat);
+    std::unique_ptr<char[]> buffer(new char[format_size]);
+    std::snprintf(buffer.get(), format_size, format.c_str(), args...);
+    return std::string(buffer.get(), buffer.get() + format_size - 1);
+}
+
```

Comment:
I feel like i've seen this function, and maybe some others here in other half PRs. If we're going to duplicate the code it might be worth adding the functino somewhere in `<root>/test_common/harness/`  and reusing it.

Reply:
@EwanC I moved it to `test_common/stringHelpers.h`